File: leet01_23.py
# 3510. Minimum Pair Removal to Sort Array II
# Solved
# Hard
# Topics
# premium lock icon
# Companies
# Hint
# Given an array nums, you can perform the following operation any number of times:

# Select the adjacent pair with the minimum sum in nums. If multiple such pairs exist, choose the leftmost one.
# Replace the pair with their sum.
# Return the minimum number of operations needed to make the array non-decreasing.

# An array is said to be non-decreasing if each element is greater than or equal to its previous element (if it exists).

 

# Example 1:

# Input: nums = [5,2,3,1]

# Output: 2

# Explanation:

# The pair (3,1) has the minimum sum of 4. After replacement, nums = [5,2,4].
# The pair (2,4) has the minimum sum of 6. After replacement, nums = [5,6].
# The array nums became non-decreasing in two operations.

# Example 2:

# Input: nums = [1,2,2]

# Output: 0

# Explanation:

# The array nums is already sorted.


# :::::::::::::::::::::::::::::  PYTHON  ::::::::::::::::::::::::::

# 🧠 Problem Recap (in plain words)

# You are given an array nums.

# You can repeatedly do one forced operation:

# 🔹 Find the adjacent pair with the minimum sum
# 🔹 If ties exist, choose the leftmost
# 🔹 Replace the pair with their sum

# Your goal is not to minimize operations arbitrarily, but to:

# 👉 Stop as soon as the array becomes non-decreasing

# That stopping condition is the most important insight.

# 🔑 Core Observations

# You do not choose which pair to merge — the problem forces the minimum-sum adjacent pair.

# An array is non-decreasing iff there is no adjacent decrease.

# Each merge only affects local neighbors.

# We need this to be fast → brute-force simulation is too slow.

# So we combine:

# Priority Queue (to always get minimum-sum pair)

# Doubly Linked List (to merge efficiently)

# Inversion counter (to know when we can stop)

# 🧩 Data Structures Used

# Let’s explain every structure and why it exists.

# 1️⃣ prev and nxt (Doubly Linked List)
# prev[i] = index before i
# nxt[i] = index after i


# Why?

# Merging deletes elements

# Python lists are slow for deletes

# Linked list gives O(1) neighbor access

# 2️⃣ alive[]
# alive[i] = whether index i still exists


# Why?

# Heap entries become outdated after merges

# We use lazy deletion

# If a pair involves a dead index → skip it

# 3️⃣ Min-Heap (heap)

# Each entry is:

# (sum_of_pair, left_index)


# Why?

# Always need the minimum adjacent sum

# Heap guarantees O(log n) access

# 4️⃣ decreaseCount
# number of adjacent pairs where nums[i] > nums[i+1]


# This is the termination condition.

# decreaseCount == 0 ⇒ array is non-decreasing

# No need to check the whole array every time

# 🧪 Step-by-Step Execution
# 🔹 Step 1: Handle Trivial Cases
# if n <= 1:
#     return 0


# A single element is always non-decreasing.

# 🔹 Step 2: Build the Linked List
# prev = [-1] * n
# nxt = [-1] * n
# alive = [True] * n


# Each element knows its neighbors.

# 🔹 Step 3: Count Initial Decreases
# decreaseCount = 0
# for i in range(n - 1):
#     if nums[i] > nums[i + 1]:
#         decreaseCount += 1


# This is crucial.

# Example:
# nums = [1, 2, 2]


# There are zero decreasing pairs → already sorted.

# ✅ Critical Early Exit
# if decreaseCount == 0:
#     return 0


# This fixes all previous wrong answers like:

# [1,2,2]

# [1,1,1]

# Without this, you would perform unnecessary merges.

# 🔹 Step 4: Initialize the Heap
# for i in range(n - 1):
#     heapq.heappush(heap, (nums[i] + nums[i + 1], i))


# Every adjacent pair is added once.

# 🔹 Step 5: Main Loop — Merge Until Sorted
# while decreaseCount > 0:


# We only keep merging while the array is still decreasing somewhere.

# 🔹 Step 6: Lazy Deletion (Very Important)
# while True:
#     s, i = heapq.heappop(heap)
#     j = nxt[i]
#     if j != -1 and alive[i] and alive[j] and nums[i] + nums[j] == s:
#         break


# Why this is needed:

# Heap contains old pairs

# Indices might be deleted

# Values might have changed

# We only accept a pair if:

# Both indices still exist

# They are still adjacent

# The sum is still correct

# 🔹 Step 7: Update decreaseCount (Before Merge)

# We remove the effects of the old relationships:

# if prev[i] != -1 and nums[prev[i]] > nums[i]:
#     decreaseCount -= 1
# if nums[i] > nums[j]:
#     decreaseCount -= 1
# if nxt[j] != -1 and nums[j] > nums[nxt[j]]:
#     decreaseCount -= 1


# Why?

# These comparisons are about to disappear

# If they were decreases, they must be removed

# 🔹 Step 8: Perform the Merge
# nums[i] += nums[j]
# alive[j] = False
# ops += 1


# We always merge right into left, exactly as the editorial requires.

# Then reconnect the linked list:

# r = nxt[j]
# nxt[i] = r
# if r != -1:
#     prev[r] = i

# 🔹 Step 9: Update decreaseCount (After Merge)

# Now new relationships are created:

# if prev[i] != -1 and nums[prev[i]] > nums[i]:
#     decreaseCount += 1
# if nxt[i] != -1 and nums[i] > nums[nxt[i]]:
#     decreaseCount += 1


# Only local neighbors can create new decreases.

# 🔹 Step 10: Push New Adjacent Pairs
# if prev[i] != -1:
#     heapq.heappush(heap, (nums[prev[i]] + nums[i], prev[i]))
# if nxt[i] != -1:
#     heapq.heappush(heap, (nums[i] + nums[nxt[i]], i))


# We add the new valid adjacent pairs caused by the merge.

# 🔹 Step 11: Termination

# When:

# decreaseCount == 0


# The array is non-decreasing — we stop immediately.

# return ops

# ✅ Why This Solution Is Correct

# ✔ Always merges the forced minimum-sum pair
# ✔ Never merges after the array becomes sorted
# ✔ Handles duplicates and negatives
# ✔ Matches the editorial logic exactly
# ✔ Passes all edge cases

# 🧠 Final Mental Model (Interview Gold)

# “We don’t choose merges.
# The problem chooses merges.
# Our job is knowing when to stop.”

import heapq

class Solution(object):
    def minimumPairRemoval(self, nums):
        n = len(nums)
        if n <= 1:
            return 0

        prev = [-1] * n
        nxt = [-1] * n
        alive = [True] * n

        for i in range(n):
            if i > 0:
                prev[i] = i - 1
            if i < n - 1:
                nxt[i] = i + 1

        # Count decreasing adjacent pairs
        decreaseCount = 0
        for i in range(n - 1):
            if nums[i] > nums[i + 1]:
                decreaseCount += 1

        # ✅ CRITICAL FIX: already sorted
        if decreaseCount == 0:
            return 0

        # Min-heap of (sum, left_index)
        heap = []
        for i in range(n - 1):
            heapq.heappush(heap, (nums[i] + nums[i + 1], i))

        ops = 0

        while decreaseCount > 0:
            # Lazy deletion
            while True:
                s, i = heapq.heappop(heap)
                j = nxt[i]
                if j != -1 and alive[i] and alive[j] and nums[i] + nums[j] == s:
                    break

            # Remove old decreases
            if prev[i] != -1 and nums[prev[i]] > nums[i]:
                decreaseCount -= 1
            if nums[i] > nums[j]:
                decreaseCount -= 1
            if nxt[j] != -1 and nums[j] > nums[nxt[j]]:
                decreaseCount -= 1

            # Merge j into i
            nums[i] += nums[j]
            alive[j] = False
            ops += 1

            r = nxt[j]
            nxt[i] = r
            if r != -1:
                prev[r] = i

            # Add new decreases
            if prev[i] != -1 and nums[prev[i]] > nums[i]:
                decreaseCount += 1
            if nxt[i] != -1 and nums[i] > nums[nxt[i]]:
                decreaseCount += 1

            # Push new adjacent pairs
            if prev[i] != -1:
                heapq.heappush(heap, (nums[prev[i]] + nums[i], prev[i]))
            if nxt[i] != -1:
                heapq.heappush(heap, (nums[i] + nums[nxt[i]], i))

        return ops

File: leet01_23.js

// "Minimum Operations to Make Array Non-Decreasing by Merging Adjacent Elements" (titled "Minimum Pair Removal to Sort Array II" in the query). This is a hard-level problem where we need to make an array nums non-decreasing (each element >= previous) by repeatedly merging the adjacent pair with the smallest sum (leftmost in case of ties), replacing them with their sum, and counting the minimum operations (merges) required.
// The code uses an efficient greedy simulation:

// Doubly Linked List: To handle merges/removals without array shifting.
// Min-Heap (Priority Queue): To always find the smallest-sum pair quickly, with tie-breaking for leftmost.
// Decrease Count: Tracks violations (adjacent pairs where left > right) to know when the array is sorted.
// Lazy Deletion: Handles stale heap entries efficiently.

// This approach ensures O(n log n) time complexity, making it suitable for typical constraints (e.g., n up to 10^5).
// The code is self-contained (no classes, just functions inside the main function) and modifies nums in place for merged sums. I'll break it down section by section, with line-by-line explanations, reasoning, and examples using nums = [5,2,3,1] (output: 2).
// Step 1: Function Signature and Edge Cases
// JavaScriptvar minimumPairRemoval = function(nums) {
//     const n = nums.length;
//     if (n <= 1) return 0;

// Explanation: Defines the function taking an array nums (assumed non-negative integers).
// n caches the length for efficiency.
// If n <= 1, no pairs exist, so it's already non-decreasing → return 0 operations.
// Why? Single element or empty array needs no merges.
// Example: [ ] or [7] → 0.

// Step 2: Setting Up the Doubly Linked List
// JavaScript// Doubly linked list simulation using indices
//     const prev = new Array(n).fill(-1);
//     const next = new Array(n).fill(-1);
//     const alive = new Array(n).fill(true);
//     for (let i = 0; i < n; i++) {
//         if (i > 0) prev[i] = i - 1;
//         if (i < n - 1) next[i] = i + 1;
//     }

// Explanation: Simulates the array as a doubly linked list via indices:
// prev[i]: Previous alive element's index (-1 if none).
// next[i]: Next alive element's index (-1 if none).
// alive[i]: True if index i is still active (not merged away).

// Loop chains them: Element 0 has prev=-1, next=1; last has next=-1.
// Why? Merging "removes" elements by skipping them (update links), avoiding O(n) array shifts. nums stores original/merged values.
// Example[5,2,3,1]:
// prev: [-1, 0, 1, 2]
// next: [1, 2, 3, -1]
// alive: [true, true, true, true]

// Effective list: 0(5) ↔ 1(2) ↔ 2(3) ↔ 3(1)

// Step 3: Counting Initial Decreasing Pairs
// JavaScript// Count number of decreasing adjacent pairs
//     let decreaseCount = 0;
//     for (let i = 0; i < n - 1; i++) {
//         if (nums[i] > nums[i + 1]) decreaseCount++;
//     }
//     if (decreaseCount === 0) return 0;

// Explanation: Scans adjacent pairs; if nums[i] > nums[i+1], it's a violation → increment decreaseCount.
// If zero, array is already non-decreasing → return 0.
// Why?decreaseCount is the loop condition (stop when 0). Updated incrementally for efficiency (no full rescan after merges).
// Example: Pairs 5>2 (yes), 2>3 (no), 3>1 (yes) → decreaseCount = 2.

// Step 4: Min-Heap Implementation (Priority Queue)
// JavaScript// ────────────────────────────────────────────────────────
//     // Min-Heap (priority queue) implementation - function style
//     // We store [sum, leftIndex]
//     // We break ties by smaller index (leftmost pair)
//     // ────────────────────────────────────────────────────────
//     const heap = [];
//     const heapPush = (item) => {
//         heap.push(item);
//         siftUp(heap.length - 1);
//     };
//     const heapPop = () => {
//         if (heap.length === 0) return null;
//         const top = heap[0];
//         const last = heap.pop();
//         if (heap.length > 0) {
//             heap[0] = last;
//             siftDown(0);
//         }
//         return top;
//     };
//     const siftUp = (i) => {
//         while (i > 0) {
//             const p = (i - 1) >> 1;
//             // Compare: smaller sum OR same sum + smaller index
//             if (compare(heap[p], heap[i]) <= 0) break;
//             [heap[p], heap[i]] = [heap[i], heap[p]];
//             i = p;
//         }
//     };
//     const siftDown = (i) => {
//         const len = heap.length;
//         while (true) {
//             let smallest = i;
//             const left = i * 2 + 1;
//             const right = i * 2 + 2;
//             if (left < len && compare(heap[left], heap[smallest]) < 0) {
//                 smallest = left;
//             }
//             if (right < len && compare(heap[right], heap[smallest]) < 0) {
//                 smallest = right;
//             }
//             if (smallest === i) break;
//             [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
//             i = smallest;
//         }
//     };
//     // Compare two entries: [sum, index]
//     // Returns negative if a < b, 0 if equal, positive if a > b
//     const compare = (a, b) => {
//         if (a[0] !== b[0]) return a[0] - b[0];
//         return a[1] - b[1]; // smaller index wins (leftmost)
//     };

// Explanation: Implements a binary min-heap storing [sum, leftIndex] for adjacent pairs.
// heap: Array holding heap elements.
// heapPush(item): Adds to end, bubbles up via siftUp (O(log n)).
// heapPop(): Removes/returns min, moves last to root, bubbles down via siftDown (O(log n)).
// siftUp/siftDown: Restore heap property using compare.
// compare(a, b): Prioritizes smaller sum; on tie, smaller index (leftmost pair).

// Why? Efficiently finds min-sum pair. Custom compare ensures problem's tie rule.
// Binary Heap Basics: Parent at (i-1)>>1, children at 2i+1/2i+2. Min-heap: parent <= children.
// Example: For [5,2,3,1], we'll push later.

// Step 5: Initializing the Heap
// JavaScript// Initialize heap with all adjacent pairs
//     for (let i = 0; i < n - 1; i++) {
//         heapPush([nums[i] + nums[i + 1], i]);
//     }

// Explanation: Pushes all initial pairs [sum, i] (i is left index).
// Why? All adjacent pairs are candidates for merging.
// Example: Push [7,0] (5+2), [5,1] (2+3), [4,2] (3+1).
// Heap after pushes (min at root): [4,2] as min (smallest sum).


// Step 6: Operations Counter
// JavaScriptlet operations = 0;

// Explanation: Tracks merge count (returned at end).

// Step 7: Main Loop - Merging Until Sorted
// JavaScriptwhile (decreaseCount > 0) {
//         let sum, i, j;
//         // Find valid minimum pair (lazy deletion)
//         while (true) {
//             const entry = heapPop();
//             if (!entry) break; // should not happen
//             [sum, i] = entry;
//             j = next[i];
//             if (j !== -1 && alive[i] && alive[j] && nums[i] + nums[j] === sum) {
//                 break;
//             }
//             // else → stale → discard
//         }

// Explanation: Loop until no decreases.
// Inner loop: Pop min [sum, i]; compute j = next[i].
// Validate: j exists, both alive, current sum matches stored (not stale post-merge).
// If invalid, discard (lazy deletion) and pop next.

// Why lazy? Merges invalidate heap entries; skipping is cheaper than removing them.
// Example (1st iter): Pop [4,2], i=2, j=3. Valid (3+1=4) → proceed.

// Step 8: Removing Old Decreasing Effects
// JavaScript// Remove effect of old decreasing relations
//         if (prev[i] !== -1 && nums[prev[i]] > nums[i]) decreaseCount--;
//         if (nums[i] > nums[j]) decreaseCount--;
//         if (next[j] !== -1 && nums[j] > nums[next[j]]) decreaseCount--;

// Explanation: Subtract violations from three old "edges":
// Left: prev[i] > i
// Middle: i > j
// Right: j > next[j]

// Why? Merge removes these; update count incrementally.
// Example: i=2(3), j=3(1), prev=1(2), next[j]=-1.
// Left: 2>3? No.
// Middle: 3>1? Yes → -- (decreaseCount=1).
// Right: Skip.


// Step 9: Performing the Merge
// JavaScript// Merge: add j into i, remove j
//         nums[i] += nums[j];
//         alive[j] = false;
//         operations++;
//         const r = next[j];
//         next[i] = r;
//         if (r !== -1) prev[r] = i;

// Explanation: Merge sum into left (i), mark j dead, increment ops.
// Update links: Skip j by setting next[i] = next[j], update prev of next.
// Why left? Arbitrary; preserves indices.
// Example: nums[2]=3+1=4, alive[3]=false, operations=1, next[2]=-1.
// Effective: [5,2,4] (indices 0,1,2 alive).

// Step 10: Adding New Decreasing Effects
// JavaScript// Add effect of new decreasing relations
//         if (prev[i] !== -1 && nums[prev[i]] > nums[i]) decreaseCount++;
//         if (next[i] !== -1 && nums[i] > nums[next[i]]) decreaseCount++;

// Explanation: Check two new edges; increment if violations.
// Example: Left: 2>4? No. Right: Skip. decreaseCount=1.

// Step 11: Pushing New Pairs
// JavaScript// Insert new possible merge candidates
//         if (prev[i] !== -1) {
//             heapPush([nums[prev[i]] + nums[i], prev[i]]);
//         }
//         if (next[i] !== -1) {
//             heapPush([nums[i] + nums[next[i]], i]);
//         }
//     }

// Explanation: Push up to two new pairs post-merge.
// Example: Push [2+4=6,1] (prev=1).
// Heap now has [5,1], [7,0], [6,1] → min [5,1] next.

// Step 12: Loop Continuation and Return

// 2nd iter: Pop [5,1], i=1, j=2. Sum 2+4=6 !=5 → stale, discard.
// Pop [6,1], valid (2+4=6).
// Remove old: 5>2? Yes → -- (decreaseCount=0), etc.
// Merge: nums[1]=2+4=6, alive[2]=false, operations=2.
// New: 5>6? No.
// Push [5+6=11,0].
// decreaseCount=0 → end.
// Return 2.

// Summary and Key Insights

// Correctness: Greedy choice (min-sum, leftmost) minimizes ops by merging smallest inversions first.
// Efficiency: Heap ops O(log n), total O(n log n).
// Edge Cases: Already sorted (0), ties (e.g., [10,1,1,10] merges 1+1 first due to leftmost)

var minimumPairRemoval = function(nums) {
    const n = nums.length;
    if (n <= 1) return 0;

    // Doubly linked list simulation using indices
    const prev = new Array(n).fill(-1);
    const next = new Array(n).fill(-1);
    const alive = new Array(n).fill(true);

    for (let i = 0; i < n; i++) {
        if (i > 0) prev[i] = i - 1;
        if (i < n - 1) next[i] = i + 1;
    }

    // Count number of decreasing adjacent pairs
    let decreaseCount = 0;
    for (let i = 0; i < n - 1; i++) {
        if (nums[i] > nums[i + 1]) decreaseCount++;
    }

    if (decreaseCount === 0) return 0;

    // ────────────────────────────────────────────────────────
    // Min-Heap (priority queue) implementation - function style
    // We store [sum, leftIndex]
    // We break ties by smaller index (leftmost pair)
    // ────────────────────────────────────────────────────────
    const heap = [];

    const heapPush = (item) => {
        heap.push(item);
        siftUp(heap.length - 1);
    };

    const heapPop = () => {
        if (heap.length === 0) return null;
        const top = heap[0];
        const last = heap.pop();
        if (heap.length > 0) {
            heap[0] = last;
            siftDown(0);
        }
        return top;
    };

    const siftUp = (i) => {
        while (i > 0) {
            const p = (i - 1) >> 1;
            // Compare: smaller sum OR same sum + smaller index
            if (compare(heap[p], heap[i]) <= 0) break;
            [heap[p], heap[i]] = [heap[i], heap[p]];
            i = p;
        }
    };

    const siftDown = (i) => {
        const len = heap.length;
        while (true) {
            let smallest = i;
            const left  = i * 2 + 1;
            const right = i * 2 + 2;

            if (left < len && compare(heap[left], heap[smallest]) < 0) {
                smallest = left;
            }
            if (right < len && compare(heap[right], heap[smallest]) < 0) {
                smallest = right;
            }
            if (smallest === i) break;

            [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
            i = smallest;
        }
    };

    // Compare two entries: [sum, index]
    // Returns negative if a < b, 0 if equal, positive if a > b
    const compare = (a, b) => {
        if (a[0] !== b[0]) return a[0] - b[0];
        return a[1] - b[1];   // smaller index wins (leftmost)
    };

    // Initialize heap with all adjacent pairs
    for (let i = 0; i < n - 1; i++) {
        heapPush([nums[i] + nums[i + 1], i]);
    }

    let operations = 0;

    while (decreaseCount > 0) {
        let sum, i, j;

        // Find valid minimum pair (lazy deletion)
        while (true) {
            const entry = heapPop();
            if (!entry) break; // should not happen

            [sum, i] = entry;
            j = next[i];

            if (j !== -1 && alive[i] && alive[j] && nums[i] + nums[j] === sum) {
                break;
            }
            // else → stale → discard
        }

        // Remove effect of old decreasing relations
        if (prev[i] !== -1 && nums[prev[i]] > nums[i]) decreaseCount--;
        if (nums[i] > nums[j]) decreaseCount--;
        if (next[j] !== -1 && nums[j] > nums[next[j]]) decreaseCount--;

        // Merge: add j into i, remove j
        nums[i] += nums[j];
        alive[j] = false;
        operations++;

        const r = next[j];
        next[i] = r;
        if (r !== -1) prev[r] = i;

        // Add effect of new decreasing relations
        if (prev[i] !== -1 && nums[prev[i]] > nums[i]) decreaseCount++;
        if (next[i] !== -1 && nums[i] > nums[next[i]]) decreaseCount++;

        // Insert new possible merge candidates
        if (prev[i] !== -1) {
            heapPush([nums[prev[i]] + nums[i], prev[i]]);
        }
        if (next[i] !== -1) {
            heapPush([nums[i] + nums[next[i]], i]);
        }
    }

    return operations;
};


